clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java

```diff
@@ -123,39 +170,117 @@ public byte magic() {
-         */
-        private Record loadRecord() throws IOException {
-            if (record != null)
-                return record;
+        @Override
+        public long pid() {
+            loadUnderlyingEntry();
+            return underlying.pid();
+        }
 
-            ByteBuffer recordBuffer = ByteBuffer.allocate(recordSize);
-            Utils.readFullyOrFail(channel, recordBuffer, position + Records.LOG_OVERHEAD, "full record");
+        @Override
+        public short epoch() {
+            loadUnderlyingEntry();
+            return underlying.epoch();
+        }
 
-            recordBuffer.rewind();
-            record = new Record(recordBuffer);
-            return record;
+        @Override
+        public int baseSequence() {
+            loadUnderlyingEntry();
+            return underlying.baseSequence();
         }
 
         @Override
-        public Record record() {
-            if (record != null)
-                return record;
+        public int lastSequence() {
+            loadUnderlyingEntry();
+            return underlying.lastSequence();
+        }
 
+        private void loadUnderlyingEntry() {
             try {
-                return loadRecord();
+                if (underlying != null)
+                    return;
+
+                ByteBuffer entryBuffer = ByteBuffer.allocate(LOG_OVERHEAD + entrySize);
+                Utils.readFullyOrFail(channel, entryBuffer, position, "full entry");
+                entryBuffer.rewind();
+
+                byte magic = entryBuffer.get(LOG_OVERHEAD + Record.MAGIC_OFFSET);
+                if (magic > LogEntry.MAGIC_VALUE_V1)
+                    underlying = new EosLogEntry(entryBuffer);
+                else
+                    underlying = new OldLogEntry.ByteBufferLogEntry(entryBuffer);
             } catch (IOException e) {
-                throw new KafkaException(e);
+                throw new KafkaException("Failed to load log entry at position " + position + " from file channel " + channel);
             }
         }
 
         @Override
+        public Iterator<LogRecord> iterator() {
+            loadUnderlyingEntry();
+            return underlying.iterator();
+        }
+
+        @Override
+        public boolean isValid() {
+            loadUnderlyingEntry();
+            return underlying.isValid();
+        }
+
+        @Override
+        public void ensureValid() {
+            loadUnderlyingEntry();
+            underlying.ensureValid();
+        }
+
+        @Override
+        public long checksum() {
+            loadUnderlyingEntry();
+            return underlying.checksum();
+        }
+
+        @Override
         public int sizeInBytes() {
-            return Records.LOG_OVERHEAD + recordSize;
+            return LOG_OVERHEAD + entrySize;
+        }
+
+        @Override
+        public void writeTo(ByteBuffer buffer) {
+            loadUnderlyingEntry();
+            underlying.writeTo(buffer);
```

Comment:
I was trying to say it seems that we (should) never do writes through FileLogInputStream? If so, perhaps the implementation can just throw UnsupportedException?